### PR TITLE
Fixed Villager Position when Sleeping

### PR DESCRIPTION
**Pull request by Savagetechguy**
This edits the position of a villager with the tag SLEEPING and will set it to the direction of the bed

This is my first Pull Request and I am also new to Java so forgive me for any mistakes

### CODE REVIEW

**Comment by rtm516** [comment]
Most of these changes should be moved to `org.geysermc.connector.entity.living.merchant.VillagerEntity`

**Comment by DoctorMacc** [comment]
It may be worth testing if there is a Java edge case where other entities sleep.

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 1]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
         setRotation(rotation);
         setOnGround(isOnGround);
         this.position = Vector3f.from(position.getX() + relX, position.getY() + relY, position.getZ() + relZ);
 
         MoveEntityAbsolutePacket moveEntityPacket = new MoveEntityAbsolutePacket();
         moveEntityPacket.setRuntimeEntityId(geyserId);
-        moveEntityPacket.setPosition(position);
-        moveEntityPacket.setRotation(getBedrockRotation());
+        //Sets Villager position and rotation when sleeping
+        if(!metadata.getFlags().getFlag(EntityFlag.SLEEPING)){
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:195]: Please use whitespace around the brackets `if (!metadata.getFlags().getFlag(EntityFlag.SLEEPING)) {`

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 2]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
+            moveEntityPacket.setPosition(position);
+            moveEntityPacket.setRotation(getBedrockRotation());
+            }else{
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:198]: Fix indentation and add whitespace `} else {`

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 3]
```diff
@@ -73,6 +76,19 @@
...
+    /**
+     * Villager sleep position
+     */
+    protected int z;
+    protected int bedId;
+    protected String bedRotationZ;
+    protected float BedPositionSubtractorW;
+    protected float BedPositionSubtractorN;
+    protected int Start;
+    protected int End;
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:88]: These should start with lowercase characters

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 4]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
+                End = bedRotationZ.indexOf("[");
+                replacement = " ";
+                Setup = bedRotationZ.substring(Start, End + 1);
+
+                switch (bedRotationZ.replace(Setup, replacement)){
+                    case " facing=south,occupied=true,part=head]":
+                    //bed is facing south
+                    z = 180;
+                    BedPositionSubtractorW = -.5f; 
+                    break;
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:210]: This should be indented

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 5]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
+                    case " facing=east,occupied=true,part=head]":
+                    //bed is facing east
+                    z = 90;
+                    BedPositionSubtractorW = -.5f;
+                    break;
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:215]: As above

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 6]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
+                    case " facing=west,occupied=true,part=head]":
+                    //bed is facing west
+                    z = 270;
+                    BedPositionSubtractorW = .5f;
+                    break;
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:220]: As above

**Comment by Savagetechguy** [comment]
I don't think any other entity have the ability to sleep. I will move the changes to the VillagerEntity file

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 7]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
...
+                    case " facing=north,occupied=true,part=head]":
+                    //rotation does not change because north is 0
+                    BedPositionSubtractorN = .5f;
+                    break;
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:224]: As above

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 8]
```diff
@@ -166,14 +182,52 @@ public void moveRelative(GeyserSession session, double relX, double relY, double
     }
 
     public void moveRelative(GeyserSession session, double relX, double relY, double relZ, Vector3f rotation, boolean isOnGround) {
+        Position bedLocation = new Position((int) position.getFloorX(), (int) position.getFloorY(), (int) position.getFloorZ());
+        bedId = session.getConnector().getWorldManager().getBlockAt(session, bedLocation);
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:186]: Chunk caching may be off - add a check to see if chunk caching is enabled: `if (session.getConnector().getConfig().isCacheChunks()`...

Comment (DoctorMacc) [reply connector/src/main/java/org/geysermc/connector/entity/Entity.java:186]: (And if it isn't you'll have to skip over your bed position checking)

Comment (Savagetechguy) [reply connector/src/main/java/org/geysermc/connector/entity/Entity.java:186]: Thanks guys I will make those changes

**Comment by Savagetechguy** [comment]
I believe that is all of the requested changes

File: connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java [hunk 9]
```diff
@@ -84,4 +92,60 @@ public void updateBedrockMetadata(EntityMetadata entityMetadata, GeyserSession s
         }
         super.updateBedrockMetadata(entityMetadata, session);
     }
+    @Override
```
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java:95]: Small nitpick - add a space before this `@Override`.

Comment (RednedEpic) [reply connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java:95]: This looks fine to me?

Comment (DoctorMacc) [reply connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java:95]: Updated - I mean they should return and then define the new function, e.g. 
```
}

@Override
```

Comment (RednedEpic) [reply connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java:96]: Gotcha. Manually resolved 🙂

File: connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java [hunk 10]
```diff
@@ -84,4 +103,56 @@ public void updateBedrockMetadata(EntityMetadata entityMetadata, GeyserSession s
...
+    public void moveRelative(GeyserSession session, double relX, double relY, double relZ, Vector3f rotation, boolean isOnGround) {
+        if (session.getConnector().getConfig().isCacheChunks()) {
+        Position bedLocation = new Position((int) position.getFloorX(), (int) position.getFloorY(), (int) position.getFloorZ());
+        bedId = session.getConnector().getWorldManager().getBlockAt(session, bedLocation);
```
[the remark this answers is not in the document]
Comment (RednedEpic) [reply connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java]: Manually resolved.

File: connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java [hunk 11]
```diff
@@ -84,4 +103,56 @@ public void updateBedrockMetadata(EntityMetadata entityMetadata, GeyserSession s
...
+        setOnGround(isOnGround);
+        this.position = Vector3f.from(position.getX() + relX, position.getY() + relY, position.getZ() + relZ);
+
+        MoveEntityAbsolutePacket moveEntityPacket = new MoveEntityAbsolutePacket();
+        moveEntityPacket.setRuntimeEntityId(geyserId);
+        //Sets Villager position and rotation when sleeping
+        if(!metadata.getFlags().getFlag(EntityFlag.SLEEPING)) {
+                moveEntityPacket.setPosition(position);
+                moveEntityPacket.setRotation(getBedrockRotation());
+            } else {
```
[the remark this answers is not in the document]
Comment (RednedEpic) [reply connector/src/main/java/org/geysermc/connector/entity/living/merchant/VillagerEntity.java]: Manually resolved.

**Comment by RednedEpic** [comment]
Manually resolved some of the spacing and indentation problems. Thanks for your contribution!

**Pull request merged**